components/sup/src/event/nats_message_stream.rs

```diff
@@ -4,93 +4,158 @@ use crate::event::{Error,
+                    }
+                    Err(e) => {
+                        Error::ConnectNatsServer(e);
+                    }
+                }
+        }
+        Ok(())
+    }
+
+    fn options_from_config(supervisor_id: &str, config: &EventStreamConfig) -> Options {
+        let name = format!("hab_client_{}", supervisor_id);
+        let ca_certs = match habitat_core::tls::native_tls_wrapper::installed_cacerts(None) {
+            Ok(ca_certs) => ca_certs,
+            Err(_)         => None
+        };
+        match config.server_certificate {
+            Some(ref nats_options) => {
+                let cert_path: PathBuf = nats_options.clone().into(); 
+                Options::with_token(&config.token.to_string())
+                    .with_name(&name)
+                    .add_root_certificate(cert_path)
+                    .add_root_certificate(ca_certs.unwrap())
+            }
+            None => {
+                Options::with_token(&config.token.to_string()) 
+                    .with_name(&name)
+                    .add_root_certificate(ca_certs.unwrap())
+            }
+        }
+    }
+
+    async fn publish(&mut self, subject: &str, msg: impl AsRef<[u8]>) -> io::Result<()> {
+        if let Some(arc) = &self.connection {
+            //let conn_guard = arc.lock().await;
+            //if let Some(conn) = &*conn_guard {
+            //let conn_guard = arc.lock().await;
+            if let Some(conn) = &*arc.lock().await {
+                conn.publish(subject, msg).await
+            }
+            else {
+                Err(io::Error::new(io::ErrorKind::Other, "Not connected to NATS server!"))
+            }
+        }
+        else {
+            Err(io::Error::new(io::ErrorKind::Other, "Not connected to NATS server!"))
+        }
+    }
+}
+
+/// NatsClient is main accessor to connect to NATS Server and
+/// to publish messages to NATS.
+#[derive(Clone)]
+struct NatsClient {
+    client: Arc<Mutex<NatsClientImpl>>,
+}
+
+impl NatsClient {
+    fn new() -> NatsClient {
+        let nats_client = NatsClientImpl::new();
+        NatsClient { client: Arc::new(Mutex::new(nats_client)) }
+    }
+
+    async fn connect(self, supervisor_id: String, config: EventStreamConfig) -> io::Result<()> {        
+        match config.connect_method.into() {
+            Some(timeout) => {
+                trace!("Timeout used -> {:?}", timeout);
+                
+                time::timeout(timeout, self.client.lock().await.connect(supervisor_id, &config))
+                    .await??;
+            } 
+            None => {
+                trace!("Timeout not used");
+                tokio::spawn(async move { self.client.lock().await.connect(supervisor_id, &config).await });
+            }
+        }
+        Ok(())
+    }
+
+    async fn publish(&mut self, subject: &str, msg: impl AsRef<[u8]>) -> io::Result<()> {
+        self.client.lock().await.publish(subject, msg).await
```

Comment:
Looks like we still have the potential for thread contention here. If there is no nats server, then the lock in line 124 will remain open possibly indefinitely. And threads attempting to publish will all wait for a lock that may never materialize.

Might consider using a try_lock here.

Reply:
The client is cloned in line 150 so I think it should be OK.